### PR TITLE
[sdk-metrics] Tags perf improvements

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Elide bounds checks in `Tags` hash code computation and equality checks
* Call `string.Equals` without passing `StringComparison` (seems to inline better)

## Benchmarks

| Method                 | NumberOfTags | Mean       | Error     | StdDev    |
|----------------------- |------------- |-----------:|----------:|----------:|
| TagsGetHashCodeOld     | 1            |  13.520 ns | 0.1686 ns | 0.1577 ns |
| TagsGetHashCodeNew     | 1            |  11.989 ns | 0.1091 ns | 0.0967 ns |
| TagsEqualsOld          | 1            |   5.802 ns | 0.0405 ns | 0.0359 ns |
| TagsEqualsNew          | 1            |   4.962 ns | 0.0696 ns | 0.0617 ns |
| TagsGetHashCodeOld     | 3            |  46.325 ns | 0.4549 ns | 0.4033 ns |
| TagsGetHashCodeNew     | 3            |  37.827 ns | 0.4984 ns | 0.3891 ns |
| TagsEqualsOld          | 3            |  14.708 ns | 0.3192 ns | 0.3278 ns |
| TagsEqualsNew          | 3            |  12.526 ns | 0.2695 ns | 0.2884 ns |
| TagsGetHashCodeOld     | 5            |  79.541 ns | 0.8260 ns | 0.7727 ns |
| TagsGetHashCodeNew     | 5            |  59.275 ns | 0.4959 ns | 0.4639 ns |
| TagsEqualsOld          | 5            |  23.636 ns | 0.2285 ns | 0.2026 ns |
| TagsEqualsNew          | 5            |  18.615 ns | 0.1292 ns | 0.1146 ns |
| TagsGetHashCodeOld     | 10           | 162.381 ns | 1.4783 ns | 1.3828 ns |
| TagsGetHashCodeNew     | 10           | 128.601 ns | 1.7367 ns | 1.6245 ns |
| TagsEqualsOld          | 10           |  45.330 ns | 0.6802 ns | 0.6362 ns |
| TagsEqualsNew          | 10           |  35.216 ns | 0.5322 ns | 0.4978 ns |

<details>
<summary>Code</summary>

```csharp
#nullable enable

using System.Diagnostics;
using System.Runtime.CompilerServices;
#if NET6_0_OR_GREATER
using System.Runtime.InteropServices;
#endif
using BenchmarkDotNet.Attributes;

namespace Benchmarks.Metrics;

public class TagsBenchmarks
{
    private KeyValuePair<string, object?>[]? tags;
    private TagsNew tagsNew;
    private TagsOld tagsOld;

    [Params(/*1, 3, 5, */10/*, 100*/)]
    public int NumberOfTags { get; set; }

    [GlobalSetup]
    public void GlobalSetup()
    {
        var tags = new KeyValuePair<string, object?>[this.NumberOfTags];

        for (int i = 0; i < this.NumberOfTags; i++)
        {
            tags[i] = new KeyValuePair<string, object?>($"tag_key_name_{i}", i);
        }

        this.tags = tags;
        this.tagsNew = new(tags);
        this.tagsOld = new(tags);
    }

    [Benchmark]
    public int TagsGetHashCodeOld()
    {
        return new TagsOld(this.tags!).GetHashCode();
    }

    [Benchmark]
    public int TagsGetHashCodeNew()
    {
        return new TagsNew(this.tags!).GetHashCode();
    }

    [Benchmark]
    public bool TagsEqualsOld()
    {
        return this.tagsOld.Equals(this.tagsOld);
    }

    [Benchmark]
    public bool TagsEqualsNew()
    {
        return this.tagsNew.Equals(this.tagsNew);
    }

    internal readonly struct TagsNew : IEquatable<TagsNew>
    {
        public static readonly TagsNew EmptyTags = new(Array.Empty<KeyValuePair<string, object?>>());

        private readonly int hashCode;

        public TagsNew(KeyValuePair<string, object?>[] keyValuePairs)
        {
            this.KeyValuePairs = keyValuePairs;
            this.hashCode = ComputeHashCode(keyValuePairs);
        }

        public readonly KeyValuePair<string, object?>[] KeyValuePairs { get; }

        public static bool operator ==(TagsNew tag1, TagsNew tag2) => tag1.Equals(tag2);

        public static bool operator !=(TagsNew tag1, TagsNew tag2) => !tag1.Equals(tag2);

        public override readonly bool Equals(object? obj)
        {
            return obj is TagsNew other && this.Equals(other);
        }

        public readonly bool Equals(TagsNew other)
        {
            var ourKvps = this.KeyValuePairs;
            var theirKvps = other.KeyValuePairs;

            if (ourKvps.Length != theirKvps.Length)
            {
                return false;
            }

#if NET6_0_OR_GREATER
            // Note: This loop uses unsafe code (pointers) to elide bounds checks on
            // two arrays we know to be of equal length.
            var cursor = ourKvps.Length;
            if (cursor > 0)
            {
                ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
                ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);
                while (true)
                {
                    // Equality check for Keys
                    if (!ours.Key.Equals(theirs.Key))
                    {
                        return false;
                    }

                    // Equality check for Values
                    if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
                    {
                        return false;
                    }

                    if (--cursor == 0)
                    {
                        break;
                    }

                    ours = ref Unsafe.Add(ref ours, 1);
                    theirs = ref Unsafe.Add(ref theirs, 1);
                }
            }
#else
            for (int i = 0; i < ourKvps.Length; i++)
            {
                ref var ours = ref ourKvps[i];

                // Note: Bounds check happens here for theirKvps element access
                ref var theirs = ref theirKvps[i];

                // Equality check for Keys
                if (!ours.Key.Equals(theirs.Key))
                {
                    return false;
                }

                // Equality check for Values
                if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
                {
                    return false;
                }
            }
#endif

            return true;
        }

        public override readonly int GetHashCode() => this.hashCode;

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int ComputeHashCode(KeyValuePair<string, object?>[] keyValuePairs)
        {
            Debug.Assert(keyValuePairs != null, "keyValuePairs was null");

#if NET6_0_OR_GREATER
            HashCode hashCode = default;

            for (int i = 0; i < keyValuePairs.Length; i++)
            {
                ref var item = ref keyValuePairs[i];
                hashCode.Add(item.Key.GetHashCode());
                hashCode.Add(item.Value);
            }

            return hashCode.ToHashCode();
#else
            var hash = 17;

            for (int i = 0; i < keyValuePairs!.Length; i++)
            {
                ref var item = ref keyValuePairs[i];
                unchecked
                {
                    hash = (hash * 31) + item.Key.GetHashCode();
                    hash = (hash * 31) + (item.Value?.GetHashCode() ?? 0);
                }
            }

            return hash;
#endif
        }
    }

    internal readonly struct TagsOld : IEquatable<TagsOld>
    {
        public static readonly TagsOld EmptyTags = new(Array.Empty<KeyValuePair<string, object?>>());

        private readonly int hashCode;

        public TagsOld(KeyValuePair<string, object?>[] keyValuePairs)
        {
            this.KeyValuePairs = keyValuePairs;

#if NET6_0_OR_GREATER
            HashCode hashCode = default;

            for (int i = 0; i < this.KeyValuePairs.Length; i++)
            {
                ref var item = ref this.KeyValuePairs[i];
                hashCode.Add(item.Key);
                hashCode.Add(item.Value);
            }

            var hash = hashCode.ToHashCode();
#else
            var hash = 17;
            for (int i = 0; i < this.KeyValuePairs.Length; i++)
            {
                ref var item = ref this.KeyValuePairs[i];
                unchecked
                {
                    hash = (hash * 31) + (item.Key?.GetHashCode() ?? 0);
                    hash = (hash * 31) + (item.Value?.GetHashCode() ?? 0);
                }
            }
#endif

            this.hashCode = hash;
        }

        public readonly KeyValuePair<string, object?>[] KeyValuePairs { get; }

        public static bool operator ==(TagsOld tag1, TagsOld tag2) => tag1.Equals(tag2);

        public static bool operator !=(TagsOld tag1, TagsOld tag2) => !tag1.Equals(tag2);

        public override readonly bool Equals(object? obj)
        {
            return obj is TagsOld other && this.Equals(other);
        }

        public readonly bool Equals(TagsOld other)
        {
            var length = this.KeyValuePairs.Length;

            if (length != other.KeyValuePairs.Length)
            {
                return false;
            }

            for (int i = 0; i < length; i++)
            {
                ref var left = ref this.KeyValuePairs[i];
                ref var right = ref other.KeyValuePairs[i];

                // Equality check for Keys
                if (!left.Key.Equals(right.Key, StringComparison.Ordinal))
                {
                    return false;
                }

                // Equality check for Values
                if (!left.Value?.Equals(right.Value) ?? right.Value != null)
                {
                    return false;
                }
            }

            return true;
        }

        public override readonly int GetHashCode() => this.hashCode;
    }
}
```
</details>

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)

### CODE REVIEW

File: src/OpenTelemetry/Metrics/AggregatorStore.cs [hunk 1]
```diff
@@ -142,6 +140,7 @@ internal sealed class AggregatorStore
         }
         else
         {
+            this.tagsToMetricPointIndexDictionary = new(concurrencyLevel: Environment.ProcessorCount, capacity: lookupCapacity);
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/AggregatorStore.cs:143]: is this related to the changes in Tags in anyway? If not, I'd suggest to keep it in a separate PR, so we can see what is the effect of that change in perf independently.

File: src/OpenTelemetry/Metrics/Tags.cs [hunk 2]
```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
...
-            ref var left = ref this.KeyValuePairs[i];
-            ref var right = ref other.KeyValuePairs[i];
+            ref var ours = ref ourKvps[i];
+
+            // Note: Bounds check happens here for theirKvps element access
+            ref var theirs = ref theirKvps[i];
 
             // Equality check for Keys
-            if (!left.Key.Equals(right.Key, StringComparison.Ordinal))
+            if (!ours.Key.Equals(theirs.Key))
```
Comment (cijothomas) [inline src/OpenTelemetry/Metrics/Tags.cs:85]: mm.. is removing Ordinal comparison *faster* ?

**Comment by cijothomas** [comment]
@CodeBlanch can you add the code used in the benchmarks for reference as well?

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:85]: It isn't removing ordinal comparison so much as it is getting to it faster 🤣

What I see before is call to `string.Equals(string, int)`:

```asm
call     [System.String:Equals(System.String,int):bool:this]
```

What I see after is inlined call to `System.SpanHelpers:SequenceEqual`:

```asm
call     [System.SpanHelpers:SequenceEqual(byref,byref,ulong):bool]
```

* [string.Equals(string) code](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L612)

* [string.Equals(string int) code](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L647C48-L647C94)

For ordinal both eventually call into [EqualsHelper](https://github.com/dotnet/runtime/blob/aea36820fdef48d6ef375508977e2f681fa4e199/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L22) which is where the `SpanHelpers.SequenceEqual` comes into play.

**Comment by CodeBlanch** [comment]
@cijothomas Added

**Comment by cijothomas** [comment]
> @cijothomas Added

Given we are ultra optimized for <=3 tags, and reasonably optimized for 3-7 tags, it'd be nicer to use a tagCount<7 in benchmarks shared, to see more practical gains.

Also, if feasible, can you show the existing benchmarks before/after, so as to get a picture of how much this affects overall perf. (hash calc and equals() is significant chunk of overall metrics cost in hot path, so I expect this to be significant. But curios if it is significant enough to show up in normal benchmark runs)

File: src/OpenTelemetry/Metrics/Tags.cs [hunk 3]
```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
...
         }
+#endif
 
         return true;
     }
 
     public override readonly int GetHashCode() => this.hashCode;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static int ComputeHashCode(KeyValuePair<string, object?>[] keyValuePairs)
```
[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:104]: Here is a diff to show the changes side-by-side...

```diff
private static int ComputeHashCode(KeyValuePair<string, object?>[] keyValuePairs)
{
        this.KeyValuePairs = keyValuePairs;

-       for (int i = 0; i < this.KeyValuePairs.Length; i++)
        for (int i = 0; i < keyValuePairs.Length; i++)
        {
-            ref var item = ref this.KeyValuePairs[i];
+            ref var item = ref keyValuePairss[i];

-            hashCode.Add(item.Key));
+            hashCode.Add(item.Key.GetHashCode()));

            hashCode.Add(item.Value);
        }
}
```

* The first change is to use the local `keyValuePairs` and not property `this.KeyValuePairs` for array access. Turns out the JIT won't elide bounds checks for array access unless it can tell for sure the reference to the array won't change during the loop. That's what using the local does.

* The second change is to call `item.Key.GetHashCode()` directly and pass in the result. This elides the `null` check in [HashCode.Add](https://github.com/dotnet/runtime/blob/9ceb2e06dd201058376d703c3822ef56d2827977/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L304) because `int` can't be `null`. [Here's a demo of what the JIT spits out.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAMyUATOS7kA3svJHKG4inIAVWRlEYorJgHMAFNQoTsghjACU+w8aBAIIAJiHOHl6+ikqBAL4BRokmlOZWuBgAkiyuVO6e3n4GsYHGoeGR3jQA4jAYABJ4ABY8ITDOPj4x8YHJyYHqqeTlADwWAHzOFuSVvv3GxaVLw2ERBTAA/DV1jbgtEG0d5Bsb5KRd80YJJUk3pQDaALJ1TQeZ/ILOzxivIe98ggA8nw2BAmLgaAA5CDZQQOByOHwAXUu5D49g8GBgQ3KzgcGBm6x8qMWy2ucSAA)

File: src/OpenTelemetry/Metrics/AggregatorStore.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:143]: It won't show on a benchmark, I don't think. I'd have to construct something whacky/custom to prove it eliminates the growth pause. Is that what are looking to see? My thinking is... we know there is growth penalty when [Concurrent]Dictionary has to expand. This change is to remove that jitter. I had a conversation with @reyang which was about people not understanding how to properly measure performance and might misinterpret that jitter as unstable SDK. Something like that, paraphrasing from memory. We do reserve the capacity when using delta reclaim today. This just does that for everything.

What I'm saying is... doesn't seem like a good use of time to split it out and try to prove in a benchmark it is useful. Ipso facto argument more or less. Happy to do it though if you _really_ want it.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/AggregatorStore.cs:143]: I didn't mean to say - we need to benchmark this particular use case. I agree with the direction of pre-allocating enough memory ahead of time. We do that wherever feasible today, and this is just filling one more gap. So I agree with this change, _but_ I always prefer to keep its own separate PR, so its clear why are we doing it, and also to show that - the gains shows in this PR is coming due to changes in Tags (hash/equals), not with pre-allocating dictionary's internal array aptly!

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/Tags.cs:85]: > For ordinal both eventually call into EqualsHelper which is where the SpanHelpers.SequenceEqual comes into play.

Are you saying that the default behavior for `string.Equals(string obj)` is to do an `Ordinal` comparison anyway? If yes, then maybe add a comment that we are doing an ordinal comparison but not explicitly specifying it to inline calls and make it faster.

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/Tags.cs:85]: interesting.. do we want to rely on .NET's implementation detail?
Or maybe the question I really wanted to ask was "Isn't `StringComparison.Ordinal` what we want to be doing? (assume no perf diff.) "

File: src/OpenTelemetry/Metrics/Tags.cs [hunk 4]
```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
...
         }
 
-        for (int i = 0; i < length; i++)
+#if NET6_0_OR_GREATER
+        // Note: This loop uses unsafe code (pointers) to elide bounds checks on
+        // two arrays we know to be of equal length.
+        var cursor = ourKvps.Length;
+        if (cursor > 0)
+        {
+            ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Tags.cs:51]: Is this and the `Unsafe.Add` calls vulnerable to garbage collector relocating this array when compacting memory?

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:85]: Yes `string.Equals(string)` does an ordinal comparison! Check out the remarks on: https://learn.microsoft.com/dotnet/api/system.string.equals?#system-string-equals(system-string)

I'll add a note on there.

Comment (utpilla) [reply src/OpenTelemetry/Metrics/Tags.cs:84]: > For ordinal both eventually call into EqualsHelper which is where the SpanHelpers.SequenceEqual comes into play.

I found this to be the case for `net6.0` and higher but it looks like for `net462` the calls to `string.Equals(string value)` and `string.Equals(string value, StringComparison.Ordinal)` lead to different code paths with the latter one having a special (possibly optimized) path for ASCII strings. Maybe remove the `StringComparison.Ordinal` parameter only for `NET6_0_OR_GREATER`.

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:51]: AFAIK they are managed pointers meaning the GC is aware of them.

I ran this pattern by @stephentoub a while back. IIRC the answer was it is safe to do so long as you know for sure the array references won't change out from under you and they are indeed the same length.

Comment (stephentoub) [reply src/OpenTelemetry/Metrics/Tags.cs:51]: If done correctly, it's valid. These are managed refs, so if the GC were to move things around, it would also update the values stored in the refs to keep them valid. The concern with code like this is it's easy to accidentally get it wrong, at which point you're subject to all the same kinds of memory safety problems as in languages like C++. As a simple for loop, the JIT would already be eliminating the bounds check on accessing the elements from whichever span was used in the for loop condition, so at best this is only eliminating the other half of the bounds checks. You'll want to be really, really sure it's worth it.

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:84]: @utpilla Interesting! Where do you see that? I just went poking around and it looks pretty similar on net framework: https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/string.cs#L614-L617

I see the special ascii code in the `StringComparison.OrdinalIgnoreCase` case is that what you were referring to?

**Comment by CodeBlanch** [comment]
@cijothomas I updated the description to show benchmarks for different number of tags. There aren't any optimizations in `struct Tags` based on length, the gains should be pretty linear based on the number of tags. It might be worth looking at if we should special case a length of 1?

Here is a run of `MetricsBenchmarks`:

Before:

| Method                                | AggregationTemporality | Mean      | Error    | StdDev   |
|-------------------------------------- |----------------------- |----------:|---------:|---------:|
| CounterHotPath                        | Cumulative             |  13.01 ns | 0.146 ns | 0.136 ns |
| CounterWith1LabelsHotPath             | Cumulative             |  53.83 ns | 0.330 ns | 0.293 ns |
| CounterWith2LabelsHotPath             | Cumulative             |  72.84 ns | 0.783 ns | 0.695 ns |
| CounterWith3LabelsHotPath             | Cumulative             |  99.68 ns | 0.652 ns | 0.578 ns |
| CounterWith4LabelsHotPath             | Cumulative             | 125.77 ns | 1.449 ns | 1.356 ns |
| CounterWith5LabelsHotPath             | Cumulative             | 144.82 ns | 1.459 ns | 1.364 ns |
| CounterWith6LabelsHotPath             | Cumulative             | 164.60 ns | 1.171 ns | 0.978 ns |
| CounterWith7LabelsHotPath             | Cumulative             | 184.49 ns | 2.803 ns | 2.485 ns |
| CounterWith1LabelsHotPathUsingTagList | Cumulative             |  64.65 ns | 1.312 ns | 1.227 ns |
| CounterWith2LabelsHotPathUsingTagList | Cumulative             |  87.66 ns | 1.096 ns | 0.972 ns |
| CounterWith3LabelsHotPathUsingTagList | Cumulative             | 114.65 ns | 1.031 ns | 0.964 ns |
| CounterWith4LabelsHotPathUsingTagList | Cumulative             | 135.71 ns | 1.352 ns | 1.265 ns |
| CounterWith5LabelsHotPathUsingTagList | Cumulative             | 156.48 ns | 1.008 ns | 0.943 ns |
| CounterWith6LabelsHotPathUsingTagList | Cumulative             | 180.64 ns | 1.736 ns | 1.624 ns |
| CounterWith7LabelsHotPathUsingTagList | Cumulative             | 202.70 ns | 2.262 ns | 2.116 ns |
| CounterWith8LabelsHotPathUsingTagList | Cumulative             | 228.39 ns | 2.670 ns | 2.498 ns |
| CounterWith9LabelsHotPathUsingTagList | Cumulative             | 294.53 ns | 2.080 ns | 1.844 ns |
| CounterHotPath                        | Delta                  |  12.62 ns | 0.039 ns | 0.034 ns |
| CounterWith1LabelsHotPath             | Delta                  |  52.23 ns | 0.441 ns | 0.413 ns |
| CounterWith2LabelsHotPath             | Delta                  |  73.28 ns | 0.727 ns | 0.680 ns |
| CounterWith3LabelsHotPath             | Delta                  |  97.32 ns | 1.266 ns | 1.184 ns |
| CounterWith4LabelsHotPath             | Delta                  | 124.98 ns | 2.041 ns | 1.909 ns |
| CounterWith5LabelsHotPath             | Delta                  | 150.81 ns | 2.569 ns | 3.685 ns |
| CounterWith6LabelsHotPath             | Delta                  | 170.89 ns | 1.589 ns | 1.408 ns |
| CounterWith7LabelsHotPath             | Delta                  | 185.56 ns | 1.646 ns | 1.540 ns |
| CounterWith1LabelsHotPathUsingTagList | Delta                  |  62.80 ns | 0.481 ns | 0.426 ns |
| CounterWith2LabelsHotPathUsingTagList | Delta                  |  89.48 ns | 0.794 ns | 0.743 ns |
| CounterWith3LabelsHotPathUsingTagList | Delta                  | 114.09 ns | 1.596 ns | 1.492 ns |
| CounterWith4LabelsHotPathUsingTagList | Delta                  | 135.33 ns | 2.359 ns | 2.206 ns |
| CounterWith5LabelsHotPathUsingTagList | Delta                  | 155.94 ns | 0.978 ns | 0.915 ns |
| CounterWith6LabelsHotPathUsingTagList | Delta                  | 180.58 ns | 0.955 ns | 0.894 ns |
| CounterWith7LabelsHotPathUsingTagList | Delta                  | 200.10 ns | 1.079 ns | 1.009 ns |
| CounterWith8LabelsHotPathUsingTagList | Delta                  | 228.79 ns | 2.795 ns | 2.615 ns |
| CounterWith9LabelsHotPathUsingTagList | Delta                  | 297.95 ns | 4.772 ns | 4.464 ns |

After:

| Method                                | AggregationTemporality | Mean      | Error    | StdDev   |
|-------------------------------------- |----------------------- |----------:|---------:|---------:|
| CounterHotPath                        | Cumulative             |  12.61 ns | 0.271 ns | 0.371 ns |
| CounterWith1LabelsHotPath             | Cumulative             |  46.16 ns | 0.480 ns | 0.449 ns |
| CounterWith2LabelsHotPath             | Cumulative             |  62.84 ns | 1.123 ns | 1.051 ns |
| CounterWith3LabelsHotPath             | Cumulative             |  92.78 ns | 1.874 ns | 4.033 ns |
| CounterWith4LabelsHotPath             | Cumulative             | 117.40 ns | 1.908 ns | 1.691 ns |
| CounterWith5LabelsHotPath             | Cumulative             | 136.79 ns | 2.619 ns | 3.671 ns |
| CounterWith6LabelsHotPath             | Cumulative             | 155.58 ns | 1.922 ns | 1.798 ns |
| CounterWith7LabelsHotPath             | Cumulative             | 171.33 ns | 2.661 ns | 2.489 ns |
| CounterWith1LabelsHotPathUsingTagList | Cumulative             |  60.23 ns | 0.851 ns | 0.796 ns |
| CounterWith2LabelsHotPathUsingTagList | Cumulative             |  79.14 ns | 1.011 ns | 0.945 ns |
| CounterWith3LabelsHotPathUsingTagList | Cumulative             | 104.30 ns | 1.196 ns | 1.119 ns |
| CounterWith4LabelsHotPathUsingTagList | Cumulative             | 122.29 ns | 1.709 ns | 1.427 ns |
| CounterWith5LabelsHotPathUsingTagList | Cumulative             | 141.39 ns | 1.343 ns | 1.256 ns |
| CounterWith6LabelsHotPathUsingTagList | Cumulative             | 163.88 ns | 1.119 ns | 1.046 ns |
| CounterWith7LabelsHotPathUsingTagList | Cumulative             | 185.83 ns | 2.857 ns | 2.672 ns |
| CounterWith8LabelsHotPathUsingTagList | Cumulative             | 209.42 ns | 1.477 ns | 1.310 ns |
| CounterWith9LabelsHotPathUsingTagList | Cumulative             | 272.17 ns | 4.786 ns | 4.477 ns |
| CounterHotPath                        | Delta                  |  12.49 ns | 0.101 ns | 0.089 ns |
| CounterWith1LabelsHotPath             | Delta                  |  45.07 ns | 0.256 ns | 0.227 ns |
| CounterWith2LabelsHotPath             | Delta                  |  63.01 ns | 0.270 ns | 0.252 ns |
| CounterWith3LabelsHotPath             | Delta                  |  86.59 ns | 1.717 ns | 1.522 ns |
| CounterWith4LabelsHotPath             | Delta                  | 116.75 ns | 2.345 ns | 4.575 ns |
| CounterWith5LabelsHotPath             | Delta                  | 131.35 ns | 2.530 ns | 2.707 ns |
| CounterWith6LabelsHotPath             | Delta                  | 150.90 ns | 2.327 ns | 1.943 ns |
| CounterWith7LabelsHotPath             | Delta                  | 173.34 ns | 3.479 ns | 5.099 ns |
| CounterWith1LabelsHotPathUsingTagList | Delta                  |  58.93 ns | 0.993 ns | 0.880 ns |
| CounterWith2LabelsHotPathUsingTagList | Delta                  |  81.29 ns | 1.625 ns | 2.577 ns |
| CounterWith3LabelsHotPathUsingTagList | Delta                  | 104.42 ns | 1.757 ns | 1.644 ns |
| CounterWith4LabelsHotPathUsingTagList | Delta                  | 126.46 ns | 2.380 ns | 2.226 ns |
| CounterWith5LabelsHotPathUsingTagList | Delta                  | 146.67 ns | 2.696 ns | 2.521 ns |
| CounterWith6LabelsHotPathUsingTagList | Delta                  | 169.62 ns | 1.926 ns | 1.801 ns |
| CounterWith7LabelsHotPathUsingTagList | Delta                  | 195.53 ns | 3.848 ns | 3.951 ns |
| CounterWith8LabelsHotPathUsingTagList | Delta                  | 205.62 ns | 2.932 ns | 2.743 ns |
| CounterWith9LabelsHotPathUsingTagList | Delta                  | 281.14 ns | 4.489 ns | 4.199 ns |

File: src/OpenTelemetry/Metrics/Tags.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/Tags.cs:84]: > I see the special ascii code in the StringComparison.OrdinalIgnoreCase case is that what you were referring to?

Ah yes. I was looking at the `OrdinalIgnoreCase` 😄

File: src/OpenTelemetry/Metrics/Tags.cs [hunk 5]
```diff
@@ -52,33 +34,100 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
...
+            ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);
+            while (true)
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Tags.cs:53]: Could you simplify this with a `for` loop?

```csharp
if (ourKvps.Length > 0)
{
    ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
    ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);

    for (int i = 0; i < ourKvps.Length; i++)
    {
        / Note: string.Equals performs an ordinal comparison
        if (!ours.Key.Equals(theirs.Key))
        {
            return false;
        }
        
        if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
        {
            return false;
        }

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
    }
}
```

Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:53]: It will be slightly slower to do that for loop style.

In the existing one we have this...

```csharp
while (true)
{
        ...

        if (--cursor == 0)
        {
             break; // Break out of loop here if done and skip the increment of pointers
        }

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
}
```

The for loop has...

```csharp
    for (int i = 0; i < ourKvps.Length; i++)
    {
        // ...

        ours = ref Unsafe.Add(ref ours, 1);
        theirs = ref Unsafe.Add(ref theirs, 1);
    }
```

It will be slightly slower because during the last iteration it will execute those two pointer increment ops for no reason.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/Metrics/Tags.cs [hunk 6]
```diff
@@ -73,7 +74,7 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
             }
         }
 #else
-        for (int i = 0; i < ourKvps.Length; i++)
+        for (int i = 0; i < length; i++)
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/Tags.cs:77]: Don't we want this to use one of the array's length properties? Would it be skipping the bounds check if we use a regular `int`?

Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/Tags.cs:77]: It does seem to elide using the local ya. Here's a demo: [SharpLab](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8DEA7ArgG30WHwFMACU7YsgWACgGABABnKYEYoBuBh8gewDM7AEzkAwv0EBvaYMEAHAE4BLAG6I4FANKkAngDVE+XKQAKiVcoA8nFjHIB7YACtSAYzgB+AHwBtAF1yAH09IxMzS2tkcgBecmxSAHdycONTCytbe0cXdy8/fw4WQN56BQV5SqYRVWw4cgBRfFUAE1IAFVJkOA4ACgBKaoUZSvGR8fJNZXIPJ1wG+PIWcqnKgDMnWf76xtVl1fIDm1D0yKyYgDoAGSoAczgAC25jgGo34Yr12Umf8mUpA200QswWyliCUBwLCBgyUWyyH8qjKfx+aPW80WjTeCXByCu8IoAEIEnhCORvOQOOQQCs1v9yABfPjff5MADscwWDQZ4yZ6zRaNqxyWLXaXR6cFEQzRY3+GMEM3IAGs4RdohDlrCIplNcg+ei2T9lVilglVoqBFsdntjodXidVeq9Yjbg9no6Pl9GeQ5MbfdCQWDcFqoUDnbqETFkaiA4yrZUzTi8aGCUTyKTEgR8JTqbT6YmWYx4+NOdzsYaFMXhXUxa0Ot1ekJZaW/YnlWqo5cw2cXdGIVWph3QeQyNhHk9ll2ifr3RPPYmR7Nkw7EzbyLslgcLY7yKdx5OvZ9E/7fQog8r8csgzONYjY0OFW3MTyU8404SLpmyTm8zS6UtF9BBrYCBHLZMn2ZaomSAA===)

Notice in "ElideTest1" there is `call 0x7ff9435f8f40` that is the thing that throws the range check failure. The other two don't have that.
